bc_obps/reporting/models/activity_json_schema.py

```diff
@@ -0,0 +1,34 @@
+from common.models import BaseModel
+from django.db import models
+from registration.models import ReportingActivity
+from reporting.models import Configuration
+import typing
+
+
+class ActivityJsonSchema(BaseModel):
+    """Intersection table for Activity-JsonSchema"""
+
+    # No history needed, these elements are immutable
+    reporting_activity = models.ForeignKey(ReportingActivity, on_delete=models.DO_NOTHING, related_name="+")
+    json_schema = models.JSONField(
+        db_comment="The json schema for a specific activity. This defines the shape of the data collected for the related activity",
+    )
+    valid_from = models.ForeignKey(Configuration, on_delete=models.DO_NOTHING, related_name="+")
+    valid_to = models.ForeignKey(Configuration, on_delete=models.DO_NOTHING, related_name="+")
+
+    class Meta:
+        db_table_comment = (
+            "Intersection table that assigns a json_schema as valid for a period of time given an activity"
+        )
+        db_table = 'erc"."activity_json_schema'
+
+    @typing.no_type_check
+    def save(self, *args, **kwargs) -> None:
+        """
+        Override the save method to validate if there are overlapping records.
+        """
+        exception_message = f'This record will result in duplicate json schemas being returned for the date range {self.valid_from.valid_from} - {self.valid_to.valid_to} as it overlaps with a current record or records'
+        from reporting.utils import validate_overlapping_records
```

Comment:
Putting the import at the top of the file creates a circular dependency as we import ActivityJsonSchema in order to type the object being passed to utils function being called just below this line

Reply:
I think this is a sign that our typing needs work.
We can probably make a base model type with the `valid_to` and `valid_from` fields, and have our model classes extend that.
The `validate_overlapping_records` function would then accept a parameter of that base model type

Reply:
something like:
```python
T = TypeVar('T', bound=ValidToFromBaseModel)

def validate_overlapping_records(object_class: T, ...):
  ...
```

Reply:
https://docs.python.org/3/library/typing.html#typing.TypeVar

Reply:
I'd like to move on from this ticket as it's a significant blocker for a lot of other work. I've set the types for the parameters to `Any` and created a tech debt ticket to apply more specific typing
https://github.com/bcgov/cas-reporting/issues/276

Reply:
I suggest bending PEP8 recommendation and using inline import here, rather than using the `Any` type, as it completely ignores the benefits of using mypy.

Reply:
Yep! Let's fix this at another time / not at all.
Also agreed with Sepehr, maybe we don't have to treat Python as a statically typed language all the time

Reply:
The function is tested in each model's test that uses it. So there is some confidence there that it is working as expected